src/OpenTK/Platform/MacOS/CocoaContext.cs

```diff
@@ -46,17 +46,52 @@ internal class CocoaContext : DesktopGraphicsContext
         private static readonly IntPtr selMakeCurrentContext = Selector.Get("makeCurrentContext");
         private static readonly IntPtr selUpdate = Selector.Get("update");
 
-        private static readonly IntPtr opengl = NS.AddImage(
+        private static IntPtr opengl = NS.AddImage(
             "/System/Library/Frameworks/OpenGL.framework/OpenGL",
             AddImageFlags.ReturnOnError);
 
-        private static readonly IntPtr opengles = NS.AddImage(
+        private static IntPtr opengles = NS.AddImage(
             "/System/Library/Frameworks/OpenGL.framework/OpenGLES",
             AddImageFlags.ReturnOnError);
+        
+        private static bool bindUsingDlSym = false;
 
         static CocoaContext()
         {
             Cocoa.Initialize();
+            DetectBindingMethod();
+        }
+
+        // As of MacOS Monterey Beta 1/2, NSLookupSymbolInImage() stopped working. It is unclear
+        // if this will be fixed by Apple in the future. dlsym still works, but is about 20x slower.
+        // We will try to detect if NSLookup works by looking up 'glClear' and fallback to dlsym 
+        // if it returns NULL.
+
+        static unsafe void DetectBindingMethod()
+        {
+            // "glClear\0"
+            var glClearName = new byte[] { 0x67, 0x6c, 0x43, 0x6c, 0x65, 0x61, 0x72, 0x00 };
+
+            fixed (byte* ptr = &glClearName[0])
+            {
+                var symbol = GetAddressUsingNSLookup(new IntPtr(ptr));
+
+                if (symbol == IntPtr.Zero)
+                {
+                    // Need to reload the library with dlopen. NSAddImage doesnt return
+                    // the same data structure. 
+                    opengl   = NS.LoadLibrary("/System/Library/Frameworks/OpenGL.framework/OpenGL");
```

Comment:
I'm guessing we want to call something like `nsRemoveImage` or whatever it's called before just reassigning these.

Reply:
I looked for such a call and to my knowledge it doesnt exist, these are the NSxxx functions:

- NSAddImage
- NSCreateObjectFileImageFromFile
- NSCreateObjectFileImageFromMemory
- NSDestroyObjectFileImage
- NSNameOfModule
- NSLibraryNameForModule
- NSLinkModule
- NSUnLinkModule
- NSIsSymbolNameDefined
- NSIsSymbolNameDefinedWithHint
- NSIsSymbolNameDefinedInImage
- NSLookupAndBindSymbol
- NSLookupAndBindSymbolWithHint
- NSLookupSymbolInImage
- NSLookupSymbolInModule
- NSNameOfSymbol
- NSAddressOfSymbol
- NSModuleForSymbol

Reply:
From dlclose documentation:

https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/dlclose.3.html

```
     Prior to Mac OS X 10.5, only bundles could be unloaded.  Starting in Mac
     OS X 10.5, dynamic libraries may also be unloaded.  There are a couple of
     cases in which a dynamic library will never be unloaded: 1) the main exe-cutable executable
     cutable links against it, 2) An API that does not supoort unloading (e.g.
     NSAddImage()) was used to load it or some other dynnamic library that
     depends on it, 3) the dynamic library is in dyld's shared cache.
```

-Mat

Reply:
I see, interesting that the api dissent allow unloading. In that case I think it would be a good idea too write a comment explaining why we are not unloading the dylib then (because there is no api for it so we can't).

Reply:
Got it, will do!!